packages/agreement-process/src/services/agreementStateProcessor.ts

```diff
@@ -144,3 +170,181 @@ export const suspendedByProducerFlag = (
   requesterOrgId === agreement.producerId
     ? destinationState === agreementState.suspended
     : agreement.suspendedByProducer;
+
+const allowedStateTransitions = (state: AgreementState): AgreementState[] =>
+  match(state)
+    .with(agreementState.draft, agreementState.pending, () => [
+      agreementState.missingCertifiedAttributes,
+    ])
+    .with(agreementState.missingCertifiedAttributes, () => [
+      agreementState.draft,
+    ])
+    .with(agreementState.active, () => [agreementState.suspended])
+    .with(agreementState.suspended, () => [
+      agreementState.active,
+      agreementState.suspended,
+    ])
+    .with(agreementState.archived, agreementState.rejected, () => [])
+    .exhaustive();
+
+const updatableStates = Object.values(agreementState).filter(
+  (state) => allowedStateTransitions(state).length > 0
+);
+
+async function updateAgreementState(
+  agreement: WithMetadata<Agreement>,
+  consumer: CompactTenant,
+  eservices: EService[],
+  correlationId: string,
+  logger: Logger
+): Promise<CreateEvent<AgreementEvent> | void> {
+  const descriptor = eservices
+    .find((eservice) => eservice.id === agreement.data.eserviceId)
+    ?.descriptors.find(
+      (descriptor) => descriptor.id === agreement.data.descriptorId
+    );
+
+  if (!descriptor) {
+    logger.error(
+      `Descriptor ${agreement.data.descriptorId} not found for Agreement ${agreement.data.id} - EService ${agreement.data.eserviceId}`
+    );
+    return;
+  }
+
+  const nextState = nextStateByAttributes(agreement.data, descriptor, consumer);
+
+  const newSuspendedByPlatform = suspendedByPlatformFlag(nextState);
+
+  const finalState = agreementStateByFlags(
+    nextState,
+    agreement.data.suspendedByProducer,
+    agreement.data.suspendedByConsumer,
+    newSuspendedByPlatform
+  );
+
+  const updatedAgreement: Agreement = {
+    ...agreement.data,
+    state: finalState,
+    suspendedByPlatform: newSuspendedByPlatform,
+  };
+
+  if (allowedStateTransitions(agreement.data.state).includes(finalState)) {
+    if (newSuspendedByPlatform !== agreement.data.suspendedByPlatform) {
+      return match(finalState)
```

Comment:
unfortunately in this case the final state is not a good discriminator
Take as example the final state = `SUSPENDED`. 
Steps
  1. the agreement is `SUSPENDED` because previously suspended by producer and platform
  2. the attribute is re-assigned and the suspension by platform is lifted
  3. the final state is still `SUSPENDED` because still suspended by producer
  4. but the persisted state is `AgreementSuspendedByPlatform` instead of `AgreementUnsuspendedByPlatform`

Reply:
Right, I missed that case! Just fixed in the last commits, and I also added a dedicated test